url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/robot/SOCPlayerTracker.java

```diff
@@ -0,0 +1,3856 @@
+    {
+        //D.ebugPrintln("%%% remove road from necessary roads");
+        Enumeration nrEnum = pr.getNecessaryRoads().elements();
+
+        while (nrEnum.hasMoreElements())
+        {
+            SOCPossibleRoad nr = (SOCPossibleRoad) nrEnum.nextElement();
+
+            //D.ebugPrintln("%%% removing road at "+Integer.toHexString(pr.getCoordinates())+" from road at "+Integer.toHexString(nr.getCoordinates()));
+            nr.getNewPossibilities().removeElement(pr);
+        }
+    }
+
+    /**
+     * remove this piece from the pieces that support it
+     *
+     * @param ps  the possible settlement
+     */
+    protected void removeFromNecessaryRoads(SOCPossibleSettlement ps)
+    {
+        if (ps == null)
+            return;    // just in case; should not happen
+
+        //D.ebugPrintln("%%% remove settlement from necessary roads");
+        Enumeration nrEnum = ps.getNecessaryRoads().elements();
+
+        while (nrEnum.hasMoreElements())
+        {
+            SOCPossibleRoad nr = (SOCPossibleRoad) nrEnum.nextElement();
+
+            //D.ebugPrintln("%%% removing settlement at "+Integer.toHexString(ps.getCoordinates())+" from road at "+Integer.toHexString(nr.getCoordinates()));
+            nr.getNewPossibilities().removeElement(ps);
+        }
+    }
+
+    /**
+     * Remove our incorrect city placement, it's been rejected by the server.
+     * Note, there is no addNewCity or addTheirNewCity method.
+     * 
+     * @param city Location of our bad city
+     * 
+     * @see SOCRobotBrain#cancelWrongPiecePlacement(SOCCancelBuildRequest)
+     */
+    public void cancelWrongCity(SOCCity city)
+    {
+        if (city == null)
+            return;      // just in case; should not happen
+
+        /**
+         * There is no addTheirNewCity method.
+         * Just remove our potential city, since it was wrongly placed.
+         * remove the possible city from the list
+         */
+        possibleCities.remove(new Integer(city.getCoordinates()));
+    }
+
+    /**
+     * add one of our cities, by removing it from the possible-cities list if it's there.
+     * Note, there is no addNewCity or addTheirNewCity method.
+     *
+     * @param city  the new city
+     */
+    public void addOurNewCity(SOCCity city)
+    {
+        /**
+         * remove the possible city from the list
+         */
+        possibleCities.remove(new Integer(city.getCoordinates()));
+    }
+
+    /**
+     * undo adding one of our cities
+     *
+     * @param city  the now possible city
+     */
+    public void undoAddOurNewCity(SOCPossibleCity city)
+    {
+        /**
+         * add the possible city to the list
+         */
+        possibleCities.put(new Integer(city.getCoordinates()), city);
+    }
+
+    /**
+     * update threats for pieces that need to be updated
+     *
+     * @param trackers  all of the player trackers
+     */
+    public void updateThreats(HashMap trackers)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 48. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 1556-1711, lines 1716-1812, lines 1817-1909 into separate methods.